### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by sofiavalenzuelapugh**
refactor TopsortBlocks to [add promote shop ](https://github.com/Topsort/building-blocks/pull/118)

### CODE REVIEW

File: demo/loader.js [hunk 1]
```diff
@@ -10,8 +10,10 @@ const isUsingTopsortBlocks = true;
 const isUsingCustomProps = false;
 const useTopsortedProducts = true;
 const customPromoteTargetClassName = "my-custom-promote-target";
-const CENTRAL_SERVICES_URL = "https://api.topsort.com/";
-const AUTH_BASE_URL = "https://buildingblocks.api.topsort.ai";
+const CENTRAL_SERVICES_URL = "http://localhost:3003";
+// const CENTRAL_SERVICES_URL = "https://api.topsort.com/";
+const AUTH_BASE_URL = "http://localhost:8000";
+// const AUTH_BASE_URL = "https://buildingblocks.api.topsort.ai";
```
Comment (PabloReszczynski) [inline demo/loader.js:16]: remove these

File: src/components/PromoteProduct.tsx [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        };
+      } else {
+        logger.warn("Missing data attributes on promote target:", {
+          "ts-product-id": productId || "(missing)",
+          "ts-product-name": productName || "(missing)",
+          "ts-product-img-url": productImgUrl || "(missing)",
+        });
+      }
+
+      return dataById;
```
Comment (PabloReszczynski) [inline src/components/PromoteProduct.tsx:66]: This can be refactored into it's own, testable function

**Review by PabloReszczynski — CHANGES_REQUESTED** [review]

**Comment by PabloReszczynski** [comment]
I'm kinda concerned about the amount of `useEffect`s happening at the same time (due to the `dispatch` dependency.
Try to decouple the business logic of the effects into their own functions that don't depend on React.

File: src/components/PromoteProduct.tsx [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+import { Portal } from "@components/Portal";
+import { usePromotionContext } from "@context";
+import { services } from "@services/central-services";
+import { Action, State } from "@state";
+import { RequestStatus } from "@types";
+import { logger } from "@utils/logger";
+import { Fragment, FunctionalComponent, h } from "preact";
+import { useEffect, useMemo, useRef, useState } from "preact/hooks";
+
+const getProductData = (promoteTargetElements: HTMLElement[]) => {
```
Comment (PabloReszczynski) [inline src/components/PromoteProduct.tsx:11]: Here, the arg is not neccesarily a HTMLElement[]. You can create an interface with the data you need

File: src/components/PromoteProduct.tsx [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+      dataById[productId] = {
+        id: productId,
+        name: productName,
+        imgUrl: productImgUrl,
+      };
+    } else {
+      logger.warn("Missing data attributes on promote target:", {
+        "ts-product-id": productId || "(missing)",
+        "ts-product-name": productName || "(missing)",
+        "ts-product-img-url": productImgUrl || "(missing)",
```
Comment (PabloReszczynski) [inline src/components/PromoteProduct.tsx:27]: Do not warn inside a reduce. Try to do the printing after the reduction is done. You can even add a new field in the accumulator to store the missing products.

**Review by PabloReszczynski — CHANGES_REQUESTED** [review]

File: src/components/PromoteProduct.tsx (hunk 3 above)
Comment (sofiavalenzuelapugh) [reply src/components/PromoteProduct.tsx:11]: it is an htlmElement[]
image

Comment (PabloReszczynski) [reply src/components/PromoteProduct.tsx:11]: What I mean is that the body of the function doesn't need the parameter to be a `HTMLElement`, just that the objects have the properties `dataset: { tsProductId: string, tsProductName: string, tsProductImgUrl: string }`

File: src/components/PromoteProduct.tsx [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+          name: productName,
+          imgUrl: productImgUrl,
+        };
+      } else {
+        missingAttributes.push({
+          "ts-product-id": productId || "(missing)",
+          "ts-product-name": productName || "(missing)",
+          "ts-product-img-url": productImgUrl || "(missing)",
+        });
+      }
```
Comment (PabloReszczynski) [inline src/components/PromoteProduct.tsx:35]: Also, don't do any kind of outside mutation inside a reduce as we want it to be a pure function. For this case, you can have the missing attributes as part of the accumulator (`dataById`).

**Review by PabloReszczynski — CHANGES_REQUESTED** [review]

**Review by PabloReszczynski — APPROVED** [review]

**Pull request merged**